REGRESSION(302776@main): Made inspector/timeline/timeline-event-TimerInstall.html flakey

dcrousso opened the pull request:
#### daecaef2cb817911209fd87b2250903593139b7e
<pre>
REGRESSION(302776@main): Made inspector/timeline/timeline-event-TimerInstall.html flakey
<a href="https://bugs.webkit.org/show_bug.cgi?id=302413">https://bugs.webkit.org/show_bug.cgi?id=302413</a>
&lt;<a href="https://rdar.apple.com/problem/164574854">rdar://problem/164574854</a>&gt;

Reviewed by BJ Burg.

Previously, there was a counter variable used to ensure that the frontend only reported that the recording was stopped once all instruments had stopped.

This didn&apos;t work when recording over a cross-origin navigation as the backend is disconnected before any messages can be sent (i.e. the frontend is told that there&apos;s a new WebProcess instead of that each instrument has stopped recording).

Ultimately, the primary purpose of that count in the frontend was to prevent the developer from starting a new recording before a previous recording had finished (assuming it existed).

The nature of WebKit&apos;s IPC means that the frontend can leverage the fact that messages are guaranteed to be delivered in the order that they&apos;re sent, so it frequently will dispatch multiple commands/events without waiting for the previous to finish.

Similarly, the frontend can wait to dispatch `WI.TimelineManager.Event.CapturingStateChanged` until any pending events from the backend have been handled.

* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager.prototype.startCapturing):
(WI.TimelineManager.prototype.stopCapturing):
(WI.TimelineManager.prototype.capturingStarted):
(WI.TimelineManager.prototype.capturingStopped):
(WI.TimelineManager.prototype._updateCapturingState): Deleted.
Include additional information about the `WI.TimelineManager.CapturingState` since `WI.TimelineManager.Event.CapturingStateChanged` is now delayed until after pending dispatches, some of which may affect the `WI.timelineManager.capturingState`.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype._handleTimelineCapturingStateChanged):
* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel.prototype._handleTimelineCapturingStateChanged):
* Source/WebInspectorUI/UserInterface/Views/TimelineOverview.js:
(WI.TimelineOverview.prototype._handleTimelineCapturingStateChanged):
* Source/WebInspectorUI/UserInterface/Views/TimelineRecordingContentView.js:
(WI.TimelineRecordingContentView.prototype._currentContentViewDidChange):
(WI.TimelineRecordingContentView.prototype._handleTimelineCapturingStateChanged):
(WI.TimelineRecordingContentView.prototype._updateProgressView):
* Source/WebInspectorUI/UserInterface/Views/TimelineRecordingProgressView.js:
(WI.TimelineRecordingProgressView.prototype.set visible):
(WI.TimelineRecordingProgressView.prototype._updateState):
(WI.TimelineRecordingProgressView.prototype._handleTimelineCapturingStateChanged):
* Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
(WI.TimelineTabContentView.prototype._updateNavigationBarButtons):
(WI.TimelineTabContentView.prototype._handleTimelineCapturingStateChanged):
(WI.TimelineTabContentView.prototype._globalModifierKeysDidChange):
(WI.TimelineTabContentView.prototype._continueButtonClicked):
Use the `WI.TimelineManager.CapturingState` included in `WI.TimelineManager.Event.CapturingStateChanged` instead of `WI.timelineManager.capturingState`.

* LayoutTests/inspector/timeline/resources/timeline-event-utilities.js:
(InspectorTest.TimelineEvent.captureTimelineWithScript):
(InspectorTest.TimelineEvent.captureTimelineWithMarker):
* LayoutTests/inspector/timeline/recording-start-stop-timestamps.html:
* LayoutTests/inspector/timeline/timeline-recording.html:

Canonical link: <a href="https://commits.webkit.org/303033@main">https://commits.webkit.org/303033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2fc801339a98e6ba0a73f9e23375e5a6372e0254

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130695 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3011 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41650 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138119 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82327 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2982 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2859 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99569 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67429 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133642 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2157 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117038 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80277 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2106 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35172 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81372 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110661 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35673 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140596 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2756 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2487 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108074 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2800 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113381 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108008 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27541 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2121 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31805 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55753 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2826 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66215 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2646 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2847 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2752 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

the-chenergy commented on Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
```diff
@@ -416,7 +422,17 @@ WI.TimelineManager = class TimelineManager extends WI.Object
 
-        this._updateCapturingState(TimelineManager.CapturingState.Active, {startTime: this._capturingStartTime});
+        this._capturingState = WI.TimelineManager.CapturingState.Active;
+        // Each instrument will send a `capturingStarted` event to the frontend so wait to notify
+        // listeners that the capturing state has changed until after receiving all of them to
+        // avoid any issues interleaving events from a previously stopped recording with commands
+        // to start a new recording (e.g. start 1a, start 1b, stop 1a, start 2, stop 1b).
+        InspectorBackend.runAfterPendingDispatches(() => {
+            this.dispatchEventToListeners(WI.TimelineManager.Event.CapturingStateChanged, {
+                capturingState: WI.TimelineManager.CapturingState.Active,
+                startTime: this._capturingStartTime,
+            });
+        });
```
I didn't quite understand what you're describing here with the interleaving events. If multiple captures may be active simultaneously, shouldn't this dispatched event also include the `activeRecording`, so that in `timeline-event-utilities.html` [(here)](https://github.com/WebKit/WebKit/pull/53827/files#diff-3fbd5919b4f5c631b5008eac2eacfb3146b27f8ed0177ba49bd5f5d512e1999aR55) we can be sure which active recording we're referring to?

dcrousso replied:
there will never be multiple timeline recordings active at the same time

consider the following series of events
1. developer asks to start recording `R1` in frontend
2. frontend sends command to backend to start timeline instrument `I1`
3. frontend sends command to backend to start timeline instrument `I2`
4. backend sends reply that timeline instrument `I1` is active
5. frontend reports that recording `R1` is active (vs in process of activating)
6. backend sends reply that timeline instrument `I2` is active
7. developer asks to stop recording `R1` in frontend
8. frontend sends command to backend to stop timeline instrument `I1`
9. frontend sends command to backend to stop timeline instrument `I2`
10. backend sends reply that timeline instrument `I1` is inactive
11. frontend reports that recording `R1` is inactive (vs in process of deactivating)
12. developer asks to start recording `R2` in frontend
13. frontend sends command to backend to start timeline instrument `I1`
14. frontend sends command to backend to start timeline instrument `I2`
15. backend sends reply that timeline instrument `I2` is inactive 🚨

so the frontend ends up in a state where although it just asked to start a timeline recording it will actually end up stopping it (when in reality that is a reply for the previous timeline recording stopping not the current one starting)

another solution would be to give each of the `trackingStart`/`trackingComplete` an identifier for which timeline recording this is, but that would not fix the issue when inspecting older backends

and to be fair this is really not possible for actual human developers since they'd not be able to click the button fast enough in step 12 (i.e. the frontend would process all the replies from the backend before the developer could ask to dispatch a new command to the backend)

the only reason why we need to do this is because a few of the tests effectively do have this behavior, so we at the very least force them to wait for all pending dispatches (represented by step 15 and thereafter) before notifying the rest of the frontend of a state change (represented by step 11)

dcrousso replied:
to more directly answer your question, we could include the `activeRecording` as part of `WI.TimelineManager.Event.CapturingStateChanged` but i dont think that'd actually be used anywhere in the UI (i.e. just in tests) so i dont think it's actually necessary

dcrousso replied:
also i think most of the time the developer manually starts a timeline recording what they're actually doing is adding new records to the existing timeline recording (i.e. Web Inspector only creates a new timeline recording if the developer presses ⇧Space or during navigations)

dcrousso committed:
REGRESSION(302776@main): Made inspector/timeline/timeline-event-TimerInstall.html flakey
https://bugs.webkit.org/show_bug.cgi?id=302413
<rdar://problem/164574854>

Reviewed by BJ Burg.

Previously, there was a counter variable used to ensure that the frontend only reported that the recording was stopped once all instruments had stopped.

This didn't work when recording over a cross-origin navigation as the backend is disconnected before any messages can be sent (i.e. the frontend is told that there's a new WebProcess instead of that each instrument has stopped recording).

Ultimately, the primary purpose of that count in the frontend was to prevent the developer from starting a new recording before a previous recording had finished (assuming it existed).

The nature of WebKit's IPC means that the frontend can leverage the fact that messages are guaranteed to be delivered in the order that they're sent, so it frequently will dispatch multiple commands/events without waiting for the previous to finish.

Similarly, the frontend can wait to dispatch `WI.TimelineManager.Event.CapturingStateChanged` until any pending events from the backend have been handled.

* Source/WebInspectorUI/UserInterface/Controllers/TimelineManager.js:
(WI.TimelineManager.prototype.startCapturing):
(WI.TimelineManager.prototype.stopCapturing):
(WI.TimelineManager.prototype.capturingStarted):
(WI.TimelineManager.prototype.capturingStopped):
(WI.TimelineManager.prototype._updateCapturingState): Deleted.
Include additional information about the `WI.TimelineManager.CapturingState` since `WI.TimelineManager.Event.CapturingStateChanged` is now delayed until after pending dispatches, some of which may affect the `WI.timelineManager.capturingState`.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager.prototype._handleTimelineCapturingStateChanged):
* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel.prototype._handleTimelineCapturingStateChanged):
* Source/WebInspectorUI/UserInterface/Views/TimelineOverview.js:
(WI.TimelineOverview.prototype._handleTimelineCapturingStateChanged):
* Source/WebInspectorUI/UserInterface/Views/TimelineRecordingContentView.js:
(WI.TimelineRecordingContentView.prototype._currentContentViewDidChange):
(WI.TimelineRecordingContentView.prototype._handleTimelineCapturingStateChanged):
(WI.TimelineRecordingContentView.prototype._updateProgressView):
* Source/WebInspectorUI/UserInterface/Views/TimelineRecordingProgressView.js:
(WI.TimelineRecordingProgressView.prototype.set visible):
(WI.TimelineRecordingProgressView.prototype._updateState):
(WI.TimelineRecordingProgressView.prototype._handleTimelineCapturingStateChanged):
* Source/WebInspectorUI/UserInterface/Views/TimelineTabContentView.js:
(WI.TimelineTabContentView.prototype._updateNavigationBarButtons):
(WI.TimelineTabContentView.prototype._handleTimelineCapturingStateChanged):
(WI.TimelineTabContentView.prototype._globalModifierKeysDidChange):
(WI.TimelineTabContentView.prototype._continueButtonClicked):
Use the `WI.TimelineManager.CapturingState` included in `WI.TimelineManager.Event.CapturingStateChanged` instead of `WI.timelineManager.capturingState`.

* LayoutTests/inspector/timeline/resources/timeline-event-utilities.js:
(InspectorTest.TimelineEvent.captureTimelineWithScript):
(InspectorTest.TimelineEvent.captureTimelineWithMarker):
* LayoutTests/inspector/timeline/recording-start-stop-timestamps.html:
* LayoutTests/inspector/timeline/timeline-recording.html:

Canonical link: https://commits.webkit.org/303033@main
